File: packages/core/src/entitlements.ts
import { OpenPayError } from "./errors";

export interface EntitlementEntry {
  userId: string;
  resourceId: string;
  grantedAt: Date;
  expiresAt: Date | null;
  /** Links the entitlement back to its originating payment for auditing. */
  paymentId?: string;
  metadata?: Record<string, string>;
}

export interface EntitlementStore {
  grant(entry: EntitlementEntry): Promise<void>;
  check(userId: string, resourceId: string): Promise<boolean>;
  revoke(userId: string, resourceId: string): Promise<void>;
  list(userId: string): Promise<EntitlementEntry[]>;
}

/** In-memory store — use in tests and demos. Not durable across restarts. */
export class MemoryEntitlementStore implements EntitlementStore {
  private readonly entries = new Map<string, EntitlementEntry>();

  private key(userId: string, resourceId: string): string {
    return `${userId}\0${resourceId}`;
  }

  async grant(entry: EntitlementEntry): Promise<void> {
    this.entries.set(this.key(entry.userId, entry.resourceId), entry);
  }

  async check(userId: string, resourceId: string): Promise<boolean> {
    const entry = this.entries.get(this.key(userId, resourceId));
    if (!entry) return false;
    if (entry.expiresAt !== null && entry.expiresAt < new Date()) {
      this.entries.delete(this.key(userId, resourceId));
      return false;
    }
    return true;
  }

  async revoke(userId: string, resourceId: string): Promise<void> {
    this.entries.delete(this.key(userId, resourceId));
  }

  async list(userId: string): Promise<EntitlementEntry[]> {
    const now = new Date();
    return Array.from(this.entries.values()).filter(
      (e) => e.userId === userId && (e.expiresAt === null || e.expiresAt > now)
    );
  }
}

export interface EntitlementManagerOptions {
  store: EntitlementStore;
}

export class EntitlementManager {
  private readonly store: EntitlementStore;

  constructor({ store }: EntitlementManagerOptions) {
    this.store = store;
  }

  async grant(params: {
    userId: string;
    resourceId: string;
    expiresAt: Date | null;
    paymentId?: string;
    metadata?: Record<string, string>;
  }): Promise<void> {
    try {
      await this.store.grant({ ...params, grantedAt: new Date() });
    } catch (err) {
      throw new OpenPayError("ENTITLEMENT_STORE_ERROR", "Failed to grant entitlement", err);
    }
  }

  async check(userId: string, resourceId: string): Promise<boolean> {
    try {
      return await this.store.check(userId, resourceId);
    } catch (err) {
      throw new OpenPayError("ENTITLEMENT_STORE_ERROR", "Failed to check entitlement", err);
    }
  }

  async revoke(userId: string, resourceId: string): Promise<void> {
    try {
      await this.store.revoke(userId, resourceId);
    } catch (err) {
      throw new OpenPayError("ENTITLEMENT_STORE_ERROR", "Failed to revoke entitlement", err);
    }
  }

  async list(userId: string): Promise<EntitlementEntry[]> {
    try {
      return await this.store.list(userId);
    } catch (err) {
      throw new OpenPayError("ENTITLEMENT_STORE_ERROR", "Failed to list entitlements", err);
    }
  }
}

File: sdk-go/entitlements.go
package openpay

import (
	"context"
	"fmt"
	"sync"
	"time"
)

// EntitlementEntry records a granted entitlement.
type EntitlementEntry struct {
	UserID     string
	ResourceID string
	GrantedAt  time.Time
	ExpiresAt  *time.Time // nil = permanent
	PaymentID  string
	Metadata   map[string]string
}

// EntitlementStore is the persistence interface for entitlements.
type EntitlementStore interface {
	Grant(ctx context.Context, entry EntitlementEntry) error
	Check(ctx context.Context, userID, resourceID string) (bool, error)
	Revoke(ctx context.Context, userID, resourceID string) error
	List(ctx context.Context, userID string) ([]EntitlementEntry, error)
}

// MemoryEntitlementStore is an in-memory EntitlementStore for tests and demos.
// Not durable across restarts.
type MemoryEntitlementStore struct {
	mu      sync.RWMutex
	entries map[string]EntitlementEntry
}

func NewMemoryEntitlementStore() *MemoryEntitlementStore {
	return &MemoryEntitlementStore{entries: make(map[string]EntitlementEntry)}
}

func (s *MemoryEntitlementStore) entryKey(userID, resourceID string) string {
	return fmt.Sprintf("%s\x00%s", userID, resourceID)
}

func (s *MemoryEntitlementStore) Grant(_ context.Context, e EntitlementEntry) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.entries[s.entryKey(e.UserID, e.ResourceID)] = e
	return nil
}

func (s *MemoryEntitlementStore) Check(_ context.Context, userID, resourceID string) (bool, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	e, ok := s.entries[s.entryKey(userID, resourceID)]
	if !ok {
		return false, nil
	}
	if e.ExpiresAt != nil && e.ExpiresAt.Before(time.Now()) {
		delete(s.entries, s.entryKey(userID, resourceID))
		return false, nil
	}
	return true, nil
}

func (s *MemoryEntitlementStore) Revoke(_ context.Context, userID, resourceID string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	delete(s.entries, s.entryKey(userID, resourceID))
	return nil
}

func (s *MemoryEntitlementStore) List(_ context.Context, userID string) ([]EntitlementEntry, error) {
	now := time.Now()
	s.mu.RLock()
	defer s.mu.RUnlock()
	var result []EntitlementEntry
	for _, e := range s.entries {
		if e.UserID == userID && (e.ExpiresAt == nil || e.ExpiresAt.After(now)) {
			result = append(result, e)
		}
	}
	return result, nil
}

// GrantInput is the input to EntitlementManager.Grant.
type GrantInput struct {
	UserID     string
	ResourceID string
	ExpiresAt  *time.Time // nil = permanent
	PaymentID  string
	Metadata   map[string]string
}

// EntitlementManager wraps an EntitlementStore with error normalisation.
type EntitlementManager struct {
	store EntitlementStore
}

func NewEntitlementManager(store EntitlementStore) *EntitlementManager {
	return &EntitlementManager{store: store}
}

func (m *EntitlementManager) Grant(ctx context.Context, input GrantInput) error {
	err := m.store.Grant(ctx, EntitlementEntry{
		UserID:     input.UserID,
		ResourceID: input.ResourceID,
		GrantedAt:  time.Now(),
		ExpiresAt:  input.ExpiresAt,
		PaymentID:  input.PaymentID,
		Metadata:   input.Metadata,
	})
	if err != nil {
		return NewError(ErrEntitlementStoreError, "failed to grant entitlement", err)
	}
	return nil
}

func (m *EntitlementManager) Check(ctx context.Context, userID, resourceID string) (bool, error) {
	ok, err := m.store.Check(ctx, userID, resourceID)
	if err != nil {
		return false, NewError(ErrEntitlementStoreError, "failed to check entitlement", err)
	}
	return ok, nil
}

func (m *EntitlementManager) Revoke(ctx context.Context, userID, resourceID string) error {
	if err := m.store.Revoke(ctx, userID, resourceID); err != nil {
		return NewError(ErrEntitlementStoreError, "failed to revoke entitlement", err)
	}
	return nil
}

func (m *EntitlementManager) List(ctx context.Context, userID string) ([]EntitlementEntry, error) {
	entries, err := m.store.List(ctx, userID)
	if err != nil {
		return nil, NewError(ErrEntitlementStoreError, "failed to list entitlements", err)
	}
	return entries, nil
}
